metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/service/OrchestratorService.java

```diff
@@ -595,46 +602,80 @@ public DatasetExecutionInformation getDatasetExecutionInformation(MetisUser meti
     final boolean isPublishCleaningOrRunning = isPluginInWorkflowCleaningOrRunning(
         runningOrInQueueExecution, PUBLISH_TYPES);
 
-    // Set the last harvest information
     final DatasetExecutionInformation executionInfo = new DatasetExecutionInformation();
-    if (lastHarvestPlugin != null) {
+    // Set the last harvest information
+    if (Objects.nonNull(lastHarvestPlugin)) {
       executionInfo.setLastHarvestedDate(lastHarvestPlugin.getFinishedDate());
       executionInfo.setLastHarvestedRecords(
           lastHarvestPlugin.getExecutionProgress().getProcessedRecords() - lastHarvestPlugin
               .getExecutionProgress().getErrors());
     }
+    final Date now = new Date();
+    setPreviewInformation(executionInfo, lastExecutablePreviewPlugin, lastPreviewPlugin,
+        isPreviewCleaningOrRunning, now);
+    setPublishInformation(executionInfo, firstPublishPlugin, lastExecutablePublishPlugin,
+        lastPublishPlugin, lastExecutableDepublishPlugin, isPublishCleaningOrRunning, now);
 
-    // Set the first publication information
-    executionInfo.setFirstPublishedDate(firstPublishPlugin == null ? null :
-        firstPublishPlugin.getFinishedDate());
+    return executionInfo;
+  }
 
+  private Date setPreviewInformation(DatasetExecutionInformation executionInfo,
+      ExecutablePlugin lastExecutablePreviewPlugin, MetisPlugin lastPreviewPlugin,
+      boolean isPreviewCleaningOrRunning, Date date) {
     // Set the last preview information
-    final Date now = new Date();
-    if (lastPreviewPlugin != null) {
+    if (Objects.nonNull(lastPreviewPlugin)) {
       executionInfo.setLastPreviewDate(lastPreviewPlugin.getFinishedDate());
       executionInfo.setLastPreviewRecordsReadyForViewing(
-          !isPreviewCleaningOrRunning && isPreviewOrPublishReadyForViewing(lastPreviewPlugin, now));
+          !isPreviewCleaningOrRunning && isPreviewOrPublishReadyForViewing(lastPreviewPlugin,
+              date));
     }
-    if (lastExecutablePreviewPlugin != null) {
+    if (Objects.nonNull(lastExecutablePreviewPlugin)) {
       executionInfo.setLastPreviewRecords(
           lastExecutablePreviewPlugin.getExecutionProgress().getProcessedRecords()
               - lastExecutablePreviewPlugin.getExecutionProgress().getErrors());
     }
+    return date;
```

Comment:
Why should this method have a return type?

Reply:
I have no idea, voided